modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/TransactionsInMultipleThreadsFailoverTest.java

```diff
@@ -0,0 +1,201 @@
+                }
+                catch (Exception ignore) {
+                    ignore.printStackTrace();
+                    assert false : "Impossible situation";
+                }
+
+                return null;
+            }
+        });
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    private void txRemoteNodeFailover() throws Exception {
+        startGrid(1);
+
+        awaitPartitionMapExchange();
+
+        IgniteCache<String, Integer> remoteCache = jcache(1);
+
+        String remotePrimaryKey = String.valueOf(primaryKey(remoteCache));
+
+        assert remotePrimaryKey != null;
+
+        try {
+            performTransactionFailover(remotePrimaryKey, 1, 0);
+
+            if (transactionConcurrency.equals(TransactionConcurrency.PESSIMISTIC))
+                fail("Broken remote node must have caused exception");
+        }
+        catch (ClusterTopologyCheckedException e) {
+            // ignoring node breakage exception
+        }
+
+        boolean txFinished = GridTestUtils.waitForCondition(new GridAbsPredicate() {
+            @Override public boolean apply() {
+                GridCacheAdapter<?, ?> cache = ((IgniteKernal)ignite(0)).internalCache(DEFAULT_CACHE_NAME);
+
+                IgniteTxManager txMgr = cache.isNear() ?
+                    ((GridNearCacheAdapter)cache).dht().context().tm() :
+                    cache.context().tm();
+
+                int txNum = txMgr.idMapSize();
+
+                return txNum == 0;
+            }
+        }, 10000);
+
+        assertTrue(txFinished);
+
+        IgniteCache<String, Integer> localCache = jcache();
+
+        if (transactionConcurrency.equals(TransactionConcurrency.OPTIMISTIC))
+            Assert.assertEquals(1, (long)localCache.get(remotePrimaryKey));
+        else
+            Assert.assertNull(localCache.get(remotePrimaryKey));
+
+        localCache.removeAll();
+    }
+
+    /**
+     * Starts tx locally with locally residing keys and then local node fails.
+     */
+    public void testTxLocalNodeFailover(){
+        withAllIsolationsAndConcurrencies(new IgniteClosure<Object, Void>() {
+            @Override public Void apply(Object o) {
+                try {
+                    txLocalNodeFailover();
+                }
+                catch (Exception ignore) {
+                    ignore.printStackTrace();
+                    assert false : "Impossible situation";
+                }
+
+                return null;
+            }
+        });
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    private void txLocalNodeFailover() throws Exception {
+        startGrid(1);
+
+        awaitPartitionMapExchange();
+
+        IgniteCache localCache = jcache(1);
+
+        String localPrimaryKey = String.valueOf(primaryKey(localCache));
+
+        assert localPrimaryKey != null;
+
+        try {
+            performTransactionFailover(localPrimaryKey, 1, 1);
+        }
+        catch (IgniteCheckedException ignore) {
```

Comment:
Let's rename "ignire" to "ignored"
Let's replace "// ignoring node breakage exception" with "// No-op."